conan/tools/system/package_manager.py

```diff
@@ -206,21 +222,33 @@ def _update(self):
             return self._conanfile_run(command, self.accepted_update_codes)
 
     def _check(self, packages, host_package=True):
-        missing = [pkg for pkg in packages if self.check_package(self.get_package_name(pkg, host_package=host_package)) != 0]
+        missing = [pkg for pkg in packages if self.check_package(pkg, host_package) != 0]
         return missing
 
-    def check_package(self, package):
-        command = self.check_command.format(tool=self.tool_name,
-                                            package=package)
+    def check_package(self, package, host_package=True):
+        name, version, arch_separator, arch_name, _ = self._split_package_name(package, host_package)
+        package = self.full_package_name.format(name=name,
+                                                arch_separator=arch_separator,
+                                                arch_name=arch_name,
+                                                version="",
+                                                version_separator="")
+        if version and self.check_version_command:
+            command = self.check_version_command.format(tool=self.tool_name, package=package, version=version)
+        else:
+            command = self.check_command.format(tool=self.tool_name, package=package)
         return self._conanfile_run(command, self.accepted_check_codes)
 
 
 class Apt(_SystemPackageManagerTool):
     # TODO: apt? apt-get?
     tool_name = "apt-get"
+    version_separator = "="
+    full_package_name = "{name}{arch_separator}{arch_name}{version_separator}{version}"
     install_command = "{sudo}{tool} install -y {recommends}{packages}"
     update_command = "{sudo}{tool} update"
-    check_command = "dpkg-query -W -f='${{Status}}' {package} | grep -q \"ok installed\""
+    check_command = "dpkg -l | grep -q -E \"^ii\\s+\\b{package}\\s+\\b\""
```

Comment:
Does the change to ``check_command`` solve issues of the previous one, or is this change more to align it with the ``check_version_command``?

Reply:
This change is necessary to fix a false positive issue we had before. It is also to align with the way of checking the version of a package.

If your package had a default architecture that was implicit and the others were explicit, the old check found the default package substring within an installation for another architecture:

For example: `foo` and `foo:armv8`. The old check said that `foo` was already installed if it found `foo:armv8`.